man/nodau.1
.TH NODAU "1" "April 2012"
.SH NAME
nodau \- simple console notetaking program
.SH SYNOPSIS
.B nodau
<OPTION> [DATA]
.SH DESCRIPTION
This manual page documents briefly the
.B nodau
command.
.PP
nodau is a simple console notetaking program.
.br
.SS "OPTIONS:"
.TP
help
print this message
.TP
list [SEARCH]
list notes, accepts optional search term
.TP
new <NAME>
create new note, <NAME> must be unique
.TP
encrypt <NAME>
encrypt a new or existing note
.TP
decrypt <NAME>
decrypt an existing note
.TP
edit <NAME>
open an existing note for editing
.TP
append <NAME>
when piping data from stdin, append it to any existing data instead of
overwritting
.TP
show <NAME>
display an existing note
.TP
del <SEARCH>
delete a note/notes, accepts name or search term
.TP
rename <OLD> <NEW>
rename an existing note
.SS "SEARCH TERM:"
.IP
search terms with spaces do not need to be inside "quotes". One exception to this convention is 
rename, in which speaces are used as delimiter between old name and new name. In such case use
"quotes" in either argument to denote spaces in note's name.
.br
<NAME>          name of a note, list will search for names
similar to the term, del will delete only an exact match
.br
t@<DATESTRING>  matches notes created at a given date/time
.br
t\-<DATESTRING>  matches notes created before a given date/time
.br
t+<DATESTRING>  matches notes created after a given date/time
.SS "DATE STRING:"
.IP
<DATESTRING> can be made of any typical date format such as:
.br
dd/mm/yy
.br
dd, mm, yyyy hh:mm
.SS "EDITING:"
.IP
By default nodau will use the EDITOR environment variable to
edit notes, simply edit the file as usual, nodau will save the
note to it's database when you exit the editor. If EDITOR is not
set the nodau builtin editor will be used.
.br
Use of the builtin editor can be forced by setting
`force_builtin_editor = true' in the nodau config file.
Alternatively, the external_editor config setting can be set to
use a specific editor regardless of the EDITOR environment
variable value. `force_builtin_editor = true' takes precedence
over EDITOR and external_editor settings.
.br
The builtin editor accepts standard printable characters, enter,
and backspace. There is no support for moving the cursor with
the arrow keys or mouse. To exit the editor and save the note,
create a new line with only a dot (.) on it, or press escape.
.SS "ENCRYPTION:"
.IP
Encrypting a note will apply DES encryption to the note using a
passphrase which is requested from the user during the process.
The passphrase is then required to decrypt, edit, or read the
note. Decrypting a note will restore the note as plain text, and
the passphrase will no longer be required for access.
.SS "CONFIGURATION FILE:"
.IP
The nodau config file is located at
$XDG_CONFIG_HOME/nodau/nodau.conf
which is typically ~/.config/nodau/nodau.conf.

The following config settings are currently in use:

SETTING                    VALUES                DEFAULT
.br
external_editor            name of a text editor executable (vim,
nano,etc).
.br
force_builtin_editor       true or false         false
.br
import_old_db              true or false         true
.br
edit_autocreate            true of false         true

If import_old_db value is true or not set, nodau will import notes
from the 0.2.x database to the 0.3.x database.
.PP
.SH AUTHOR
This manual page was modified by Lisa Milne <lisa@ltmnet.com> from
an original written by Salvatore Bonaccorso <carnil@debian.org>
for the Debian system (but may be used by others).



CHANGELOG
CHANGES:
0.1
	- initial release
0.2
	- added 'show' option, to display a note without editing
	- changed 'open' option to 'edit' to differentiate it from
	 'show', 'open' has been left for backward compatibility
0.2.1
	- added man page thanks to Salvatore Bonaccorso
	 <carnil <at> debian.org>
0.2.2
	- patch to build/Makefile to fix bug when building on ubuntu
	 provided by Angel Abad
	 http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=627306
	 https://bugs.launchpad.net/ubuntu/+source/nodau/+bug/785135
0.2.3
	- patches to Makefiles to improve build portability provided by
	 Mats Andersson <bsd <at> gisladisker.se>
0.3rc1
	- use an external editor for editing using the EDITOR
	 environment variable
	- move database location to comply with XDG standard
	- rewritten Makefiles
	- add support for encrypted notes
	- add a config file
0.3rc2
	- add GPL exception for linking with OpenSSL
	- updates to Makefile to allow passing of flags by environment
	 variable
	- updates to man page about configuration file by Salvatore
	 Bonaccorso <carnil <at> debian.org>
	- updates to man page and README file to document configuration
	 settings by Ryan MacNish <ryan <at> nisshh.com>
0.3rc3
	- only give user write permission when creating config/data
	 directories
	- prevent segfault when list'ing with no matches
0.3rc4
	- remove dependency on PATH_MAX to hopefully improve GNU Hurd
	 compatability
0.3rc5
	- remove dependency on PATH_MAX to hopefully improve GNU Hurd
	 compatability - do it properly!
0.3rc6
	- fix a segv in the database code
	- make encryption work again (when did this get broken?)
0.3rc7
	- allow passing CPPFLAGS by Salvatore Bonaccorso
	 <carnil <at> debian.org>
0.3.0
	- no changes, it seems stable
0.3.1
	- fix various compiler warnings
0.3.2
	- add 'create' as a synonym for 'new'
	- make 'edit' auto-create a note if it does not exist and
	 edit_autocreate is true in config
0.3.3
	- some documentation improvements by Salvatore Bonaccorso
	 <carnil <at> debian.org>
0.3.4
	- fixes to Makefile by Salvatore Bonaccorso
	 <carnil <at> debian.org>
0.3.5
	- make note editing scriptable, by allowing note content to be
	 piped to stdin
	- stop stack corruption on db_update with long notes
	- changes to temporary file handling to fix security
	 vulnerability - https://github.com/darkrose/nodau/issues/17
0.3.6
	- add append option to help text by Salvatore Bonaccorso
	 <carnil <at> debian.org>
	- fix compiler warning in usage()
0.3.7
	- prevent segv when env is empty
	- prevent segv in config file creation
0.3.8
	- add support for openssl 1.1 (by Salvatore Bonaccorso
	 <carnil <at> debian.org>)
	- fix typo in an error message
	- a bunch of small fixes to remove some compiler warnings
0.3.9
	- fix segfault when running 'nodau list' with search keyword,
	 and there are no matches (by Omer Dagan
	 <mr.omer.dagan <at> gmail.com>)
	- remove variable definitions in header file that caused issues
	 with gcc 10
0.3.10
    - Move to parametetrized SQL statement allows using chars like "'"
     in notes, and also avoid sql injection
0.3.11
    - add 'rename' option to allow renaming of existing notes by Omer
    Dagan <mr.omer.dagan <at> gmail.com>
0.3.12
    - remove harcoded limit of note text size in function db_insert
    by Omer Dagan <mr.omer.dagan <at> gmail.com>


src/nodau.c
/************************************************************************
* nodau.c
* nodau console note taker
* Copyright (C) Lisa Milne 2010-2013 <lisa@ltmnet.com>
*
* nodau.c is free software: you can redistribute it and/or modify it
* under the terms of the GNU General Public License as published by the
* Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* nodau.c is distributed in the hope that it will be useful, but
* WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
* See the GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>
*
* In addition, as a special exception, the copyright holder (Lisa Milne)
* gives permission to link the code of this release of nodau with the
* OpenSSL project's "OpenSSL" library (or with modified versions of it
* that use the same license as the "OpenSSL" library), and distribute
* the linked executables. You must obey the GNU General Public License
* in all respects for all of the code used other than "OpenSSL". If you
* modify this file, you may extend this exception to your version of the
* file, but you are not obligated to do so. If you do not wish to do so,
* delete this exception statement from your version.
************************************************************************/

#include "nodau.h"

/* print help and usage */
static void usage()
{
	printf(
		"%s %s - simple console notetaking program\n\n"
		"USAGE: %s <OPTION> [DATA]\n\n"
		"OPTIONS:\n"
		" help                print this message\n"
		" list [search]       list notes, accepts optional search term\n"
		" new <name>          create new note, name must be unique\n"
		" encrypt <name>      encrypt a new or existing note\n"
		" decrypt <name>      decrypt an encrypted note\n",
		TARGET,
		VERSION,
		TARGET
	);
	printf(
		" edit <name>         open an existing note for editing\n"
		" append <name>       when piping data from stdin, append to an existing note\n"
		" show <name>         display an existing note\n"
		" del <search>        accepts name or search term\n"
		" rename <old> <new>  rename an existing note\n\n"
		"See the nodau man page for more details.\n\n"
	);

}

/* assemble arguments into a single argument, undoes the space splitting */
static char* get_args(int argc, char** argv)
{
	int i;
	char* args;
	int l = 0;

	/* get the compined length of the arguments */
	for (i=2; i<argc; i++) {
		l += strlen(argv[i]);
	}

	/* no arguments, return null */
	if (l == 0)
		return NULL;

	/* add spaces to the argument length */
	l += i;

	/* create space for the argument */
	args = malloc(l);
	/* if null throw an error */
	if (args == NULL) {
		fprintf(stderr,"an error occurred in argument compilation\n");
		return NULL;
	}

	/* set up for joining */
	strcpy(args,argv[2]);

	/* join the arguments */
	for (i=3; i<argc; i++) {
		strcat(args," ");
		strcat(args,argv[i]);
	}

	/* return the string */
	return args;
}

int main(int argc, char** argv)
{
	char* args;
	/* no option, print usage */
	if (argc < 2) {
		usage();
		return 0;
	}

	config_load();

	/* connect to the db or error */
	if (db_connect()) {
		fprintf(stderr, "Can't open database: %s\n", db_err());
		db_close();
		return 0;
	}

	/* compile the arguments */
	args = get_args(argc,argv);

	/* if listing notes */
	if (strcmp(argv[1],"list") == 0) {
		db_list(args);
	}else{
		/* if null argument print usage */
		if (args == NULL || argc <3) {
			usage();
		/* if creating a new note */
		}else if (strcmp(argv[1],"new") == 0 || strcmp(argv[1],"create") == 0) {
			db_new(args);
		/* if opening/editing an existing note */
		}else if (strcmp(argv[1],"open") == 0 || strcmp(argv[1],"edit") == 0) {
			db_edit(args);
		/* append to a note if data is on stdin */
		}else if (strcmp(argv[1],"append") == 0) {
			if (isatty(STDIN_FILENO)) {
				db_edit(args);
			}else{
				db_append(args);
			}
		/* encrypt a new or existing note */
		}else if (strcmp(argv[1],"encrypt") == 0) {
			db_encrypt(args);
		/* decrypt an existing note */
		}else if (strcmp(argv[1],"decrypt") == 0) {
			db_decrypt(args);
		/* display an existing note */
		}else if (strcmp(argv[1],"show") == 0) {
			db_show(args);
		/* if deleting note/s */
		}else if (strcmp(argv[1],"del") == 0) {
			db_del(args);
		/* rename note */
		}else if (strcmp(argv[1],"rename") == 0) {
			db_rename(argv[2], argv[3]);
		/* unknown option, print usage */
		}else{
			usage();
		}
	}

	/* free args if we can */
	if (args != NULL)
		free(args);

	/* close the database */
	db_close();

	/* save config */
	config_save();

	return 0;
}
